Sources/SmithyCodegenCore/AST/ASTNode.swift

```diff
@@ -0,0 +1,45 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/// Contains the value of a Smithy Node, as used in a JSON AST.
+///
+/// Smithy node data is basically the same as the data that can be stored in JSON.
+/// The root of a Smithy node may be of any type, i.e. unlike JSON, the root element is not limited to object or list.
+///
+/// See the definition of node value in the Smithy spec: https://smithy.io/2.0/spec/model.html#node-values
+enum ASTNode {
+    case object([String: ASTNode])
+    case list([ASTNode])
+    case string(String)
+    case number(Double)
+    case boolean(Bool)
+    case null
+}
+
```

Comment:
Because a node in AST can be of multiple JSON types, the hand-written `Decodable` implementation below is provided to capture any one of them.